src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java

```diff
@@ -84,9 +84,9 @@ ByteBuffer makeByteBuffer() {
 
     // factories
 
-    public static class OfByte extends HeapMemorySegmentImpl<byte[]> {
+    public static class OfByte extends HeapMemorySegmentImpl {
 
-        OfByte(long offset, byte[] base, long length, int mask) {
+        OfByte(long offset, Object base, long length, int mask) {
```

Comment:
Is it possible to retain the array type for the constructor?

Reply:
yes, that should be harmless

Reply:
Actually, no, sorry. If I sharpen the type, then `dup` has to cast, so we go back to the original issue.

Reply:
Ah yes, i see. Maybe it's worth adding a general comment to the class doc of `HeapMemorySegmentImpl` briefly explaining the reasons for use of `Object`.